app/src/main/cpp/skyline/gpu/texture/layout.cpp

```diff
@@ -89,144 +89,428 @@ namespace skyline::gpu::texture {
                             else
                                 std::memcpy(sector, linearSector + pixelOffset, formatBpb);
+                        }
 
                         sector += formatBpb;
                         xT += formatBpb;
                     }
                 });
         }};
 
-        u8 *linearRob{linear};
+        u8 *pitchRob{pitch};
         for (size_t rob{}; rob < surfaceHeightRobs; rob++) { // Every Surface contains `surfaceHeightRobs` ROBs (excl. padding ROB)
-            deswizzleRob(linearRob, std::false_type{});
-            linearRob += robBytes; // Increment the linear ROB to the next ROB
+            deswizzleRob(pitchRob, std::false_type{});
+            pitchRob += robBytes; // Increment the linear ROB to the next ROB
         }
 
         if (surfaceHeightLines % robHeight != 0) {
             blockHeight = (util::AlignUp(surfaceHeightLines, GobHeight) - (surfaceHeightRobs * robHeight)) / GobHeight; // Calculate the amount of Y GOBs which aren't padding
 
-            size_t alignedSurfaceLines{util::DivideCeil<size_t>(dimensions.height, formatBlockHeight)};
             deswizzleRob(
-                linearRob,
+                pitchRob,
                 std::true_type{},
-                (gobBlockHeight - blockHeight) * (SectorWidth * SectorWidth * SectorHeight), // Calculate padding at the end of a block to skip
-                util::IsAligned(alignedSurfaceLines, GobHeight) ? GobHeight : alignedSurfaceLines - util::AlignDown(alignedSurfaceLines, GobHeight) // Calculate the line relative to the start of the last GOB that is the cut-off point for the image
+                (gobBlockHeight - blockHeight) * (GobWidth * GobHeight), // Calculate padding at the end of a block to skip
+                util::IsAligned(surfaceHeightLines, GobHeight) ? GobHeight : surfaceHeightLines - util::AlignDown(surfaceHeightLines, GobHeight) // Calculate the line relative to the start of the last GOB that is the cut-off point for the image
+            );
+        }
+    }
+
+    /**
+     * @brief Copies pixel data between a pitch and part of a blocklinear texture
+     * @tparam BlockLinearToPitch Whether to copy from a part of a blocklinear texture to a pitch texture or a pitch texture to a part of a blocklinear texture
+     * @note The function assumes that the pitch texture is always equal or smaller than the blocklinear texture
+     */
+    template<bool BlockLinearToPitch>
+    void CopyBlockLinearSubrectInternal(const Dimensions &pitchDimensions, const Dimensions &blockLinearDimensions,
+                                 size_t formatBlockWidth, size_t formatBlockHeight, size_t formatBpb, u32 pitchAmount,
+                                 size_t gobBlockHeight, size_t gobBlockDepth,
+                                 u8 *blockLinear, u8 *pitch,
+                                 u16 originX, u16 originY) {
+        const u32 blockSize{(u32)(GobWidth * GobHeight * gobBlockHeight * gobBlockDepth)};
```

Comment:
Aside from all the `const`s and C-style casts, I feel this function is a bit too overcomplicated in the approach taken to subrect deswizzling so it'll likely be replaced. The terminology used is also really iffy but it's best to not bother changing much since I'll likely rewrite it as a whole.